### PR TITLE
Remote configurable Life feed URL

### LINKED ISSUE #3350: Remote configurable Life feed URL
**User story**
As life feed service provider, I would like to be able to use Firebase as remote config backend to configure service URL (API)

** acceptance criteria **
1. Supports versioning of the API URL to map to corresponding client app

### PR DESCRIPTION
**Pull request by benjamin-cheng**
Close #3350

### CODE REVIEW

File: app/src/main/java/org/mozilla/focus/utils/AppConfigWrapper.java [hunk 1]
```diff
@@ -101,6 +106,26 @@ public static boolean isLifeFeedEnabled(Context context) {
...
+            JSONArray rows = new JSONArray(source);
+            for (int i = 0; i < rows.length(); i++) {
+                JSONObject row = rows.getJSONObject(i);
+                if (row.getString("name").equalsIgnoreCase(provider)) {
+                    url = row.getString("url");
+                    break;
+                }
+            }
+        } catch (JSONException e) {
+            e.printStackTrace();
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/focus/utils/AppConfigWrapper.java]: I think you should return the default value in this case

File: app/src/main/java/org/mozilla/focus/utils/FirebaseHelper.java [hunk 2]
```diff
@@ -219,6 +220,9 @@ protected Void doInBackground(Void... voids) {
                         settings.setNewsSource(source);
                         NewsSourceManager.getInstance().setNewsSource(source);
                     }
+
+                    final String url = AppConfigWrapper.getLifeFeedProviderUrl(applicationContext, settings.getNewsSource());
+                    ThreadUtils.postToMainThread(() -> NewsSourceManager.getInstance().setNewsSourceUrl(url));
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/focus/utils/FirebaseHelper.java]: why post to main thread?

File: app/src/main/java/org/mozilla/rocket/content/NewsSourceManager.java [hunk 3]
```diff
@@ -67,6 +73,15 @@ public void setNewsSource(String newsSource) {
         ContentRepository.reset();
     }
 
+    public String getNewsSourceUrl() {
+        return newsSourceUrl;
+    }
+
+    public void setNewsSourceUrl(String newsSourceUrl) {
+        this.newsSourceUrl = newsSourceUrl;
+        ContentRepository.resetSubscriptionUrl(newsSourceUrl);
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/rocket/content/NewsSourceManager.java]: why not reuse reset()?

File: components/partners/partnerrepository/src/main/java/org/mozilla/lite/partner/Repository.java [hunk 4]
```diff
@@ -50,17 +52,23 @@ public void loadMore() {
         nextSubscription();
     }
 
-    // Not used yet and not tested yet.
     public void reloadData() {
         itemPojoList = new ArrayList<>();
         cacheIsDirty = true;
+        currentPage = firstPage;
```
Comment (cnevinc) [inline components/partners/partnerrepository/src/main/java/org/mozilla/lite/partner/Repository.java]: can you call reset and then ` nextSubscription();` ?

File: app/src/main/java/org/mozilla/rocket/content/ContentRepository.kt [hunk 5]
```diff
@@ -32,14 +31,19 @@ class ContentRepository {
...
+            INSTANCE?.setSubscriptionUrl(subscriptionUrl)
+        }
+
         @JvmStatic
         fun isEmpty() = INSTANCE == null
 
         private fun buildRepository(context: Context): Repository<out NewsItem> {
             return if (NewsSourceManager.getInstance().newsSource == NEWS_DB) {
-                RepositoryBhaskar(context)
+                RepositoryBhaskar(context, NewsSourceManager.getInstance().newsSourceUrl)
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/rocket/content/ContentRepository.kt]: Can we derive the URL from the current source and it's URL in AppConstant? I'll like to remove the NewsSourceManager in the future cause it lacks testing and make the URL/Source hard to sync.

File: app/src/main/java/org/mozilla/focus/utils/AppConfigWrapper.java (hunk 1 above)
Comment (benjamin-cheng) [reply app/src/main/java/org/mozilla/focus/utils/AppConfigWrapper.java]: Empty url may be better than broken(default) url

File: app/src/main/java/org/mozilla/focus/utils/FirebaseHelper.java (hunk 2 above)
Comment (benjamin-cheng) [reply app/src/main/java/org/mozilla/focus/utils/FirebaseHelper.java]: setNewsSourceUrl() will trigger series actions related to UI update.
setNewsSourceUrl() -> Repository.nextSubscription() -> onDataChangedListener::onDataChanged() -> ContentViewModel.item.setValue()

File: app/src/main/java/org/mozilla/rocket/content/NewsSourceManager.java (hunk 3 above)
Comment (benjamin-cheng) [reply app/src/main/java/org/mozilla/rocket/content/NewsSourceManager.java]: For this case, it should not set `onDataChangedListener` as null in reset()

File: components/partners/partnerrepository/src/main/java/org/mozilla/lite/partner/Repository.java (hunk 4 above)
Comment (benjamin-cheng) [reply components/partners/partnerrepository/src/main/java/org/mozilla/lite/partner/Repository.java]: For this case, it should not set onDataChangedListener as null in reset()

File: app/src/main/java/org/mozilla/rocket/content/ContentRepository.kt (hunk 5 above)
Comment (benjamin-cheng) [reply app/src/main/java/org/mozilla/rocket/content/ContentRepository.kt]: Maybe we could refine the class responsibility when we try to remove NewsSourceManager. Currently, it should be reasonable to keep url in NewsSourceManager.

**Comment by cnevinc** [comment]
Could you rename the commit to align below style?
Closes #3350 - Your commit message.

File: app/src/main/java/org/mozilla/focus/utils/FirebaseHelper.java (hunk 2 above)
Comment (cnevinc) [reply app/src/main/java/org/mozilla/focus/utils/FirebaseHelper.java]: Please correct me if I'm wrong. 
Repository.nextSubscription() -> Could run in background thread
onDataChangedListener::onDataChanged() -> it's called in LiveData's onChanged so it's UI thread.
ContentViewModel.item.setValue() -> same as above

**Comment by benjamin-cheng** [comment]
OK, rename and squash the commits, thanks for the review 💯

**Pull request merged**